python/ray/train/_internal/storage.py

```diff
@@ -0,0 +1,154 @@
+
+        if storage_filesystem:
+            # Custom pyarrow filesystem
+            self.storage_filesystem = storage_filesystem
+            if is_uri(self.storage_path):
+                raise ValueError("TODO")
+            self.storage_fs_path = self.storage_path
+        else:
+            (
+                self.storage_filesystem,
+                self.storage_fs_path,
+            ) = get_fs_and_path(self.storage_path)
+
+        self.syncer: Optional[Syncer] = (
+            None
+            if self.storage_path == self.storage_cache_path
+            else _DefaultSyncer(
+                sync_period=self.sync_config.sync_period,
+                sync_timeout=self.sync_config.sync_timeout,
+                storage_filesystem=self.storage_filesystem,
+            )
+        )
+
+        # TODO(justinvyu): This is needed for the legacy Trainable syncing code.
+        # Remove this after syncer argument is removed from SyncConfig.
+        self.sync_config.syncer = self.syncer
+
+        self._create_validation_file()
+        self._check_validation_file()
+
+    def __str__(self):
+        attrs = [
+            "storage_path",
+            "storage_cache_path",
+            "storage_filesystem",
+            "storage_fs_path",
+            "experiment_dir_name",
+            "trial_dir_name",
+        ]
+        attr_str = "\n".join([f"  {attr}={getattr(self, attr)}" for attr in attrs])
+        return f"StorageContext<\n{attr_str}\n>"
+
+    def _create_validation_file(self):
+        valid_file = os.path.join(self.experiment_fs_path, ".validate_storage_marker")
+        self.storage_filesystem.create_dir(self.experiment_fs_path)
+        with self.storage_filesystem.open_output_stream(valid_file):
+            pass
+
+    def _check_validation_file(self):
+        valid_file = os.path.join(self.experiment_fs_path, ".validate_storage_marker")
+        valid = self.storage_filesystem.get_file_info([valid_file])[0]
+        if valid.type == pyarrow.fs.FileType.NotFound:
+            raise RuntimeError(
+                f"Unable to set up cluster storage at storage_path={self.storage_path}"
+                "\nCheck that all nodes in the cluster have read/write access "
+                "to the configured storage path."
+            )
+
+    @property
+    def experiment_fs_path(self) -> str:
+        return os.path.join(self.storage_fs_path, self.experiment_dir_name)
+
+    @property
+    def experiment_cache_path(self) -> str:
+        return os.path.join(self.storage_cache_path, self.experiment_dir_name)
+
+    @property
+    def trial_fs_path(self) -> str:
+        if self.trial_dir_name is None:
+            raise RuntimeError(
+                "Should not access `trial_fs_path` without setting `trial_dir_name`"
+            )
+        return os.path.join(self.experiment_fs_path, self.trial_dir_name)
+
+    def construct_checkpoint_fs_path(self, checkpoint_dir_name: str) -> str:
```

Comment:
Similar to trial_fs_path, let's make this a property that is generated based on current checkpoint number + the trial fs path, instead of allowing checkpoint_dir_name to be passed on as a method argument.